src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceVariableService.cs

```diff
@@ -293,15 +300,18 @@ private CodeAction CreateAction(State state, bool allOccurrences, bool isConstan
             SemanticDocument currentDocument,
             SyntaxNode withinNodeInCurrent,
             bool allOccurrences,
+            bool includeRValues,
             CancellationToken cancellationToken)
         {
             var syntaxFacts = currentDocument.Project.LanguageServices.GetService<ISyntaxFactsService>();
+            var semanticFacts = currentDocument.Project.LanguageServices.GetService<ISemanticFactsService>();
             var originalSemanticModel = originalDocument.SemanticModel;
             var currentSemanticModel = currentDocument.SemanticModel;
 
             var result = new HashSet<TExpressionSyntax>();
             var matches = from nodeInCurrent in withinNodeInCurrent.DescendantNodesAndSelf().OfType<TExpressionSyntax>()
                           where NodeMatchesExpression(originalSemanticModel, currentSemanticModel, expressionInOriginal, nodeInCurrent, allOccurrences, cancellationToken)
+                            && (includeRValues || (!includeRValues && !allOccurrences) || (!includeRValues && allOccurrences && semanticFacts.CanReplaceWithRValue(currentSemanticModel, nodeInCurrent, cancellationToken)))
```

Comment:
either extract this to a local function or re-format the expression to be on multiple lines

Reply:
I split them on multiple lines and added detailed comments explaining each of the 3 parts of that condition